File: scarl/src/main/scala/io/github/fiifoo/scarl/world/system/Physics.scala
package io.github.fiifoo.scarl.world.system

object Physics {

  val g: Double = 9.80665

  val G: Double = 6.67408 * math.pow(10, -11)

  val AU: Double = 149597871000.0

  val EarthMass: Double = 5.97237 * math.pow(10, 24)
  val SolarMass: Double = 1.98847 * math.pow(10, 30)

  def hypotenuse(a: Double, b: Double): Double = math.sqrt(math.pow(a, 2) + math.pow(b, 2))

  def vector(slope: Double, length: Double): Vector = {
    val x = length / math.sqrt(1 + math.pow(slope, 2))
    val y = slope * x

    Vector(x, y)
  }

  def travelTime(distance: Double, acceleration: Double): Int = math.sqrt(2 * distance / acceleration).toInt

  def gravity(bodies: Iterable[StellarBody], body: StellarBody): Vector = {
    bodies filter (_.id != body.id) map calculateGravity(body) reduceLeft (_ add _)
  }

  private def calculateGravity(body1: StellarBody)(body2: StellarBody): Vector = {
    val quarter = Coordinates.Quarter(body1.position, body2.position)
    val p1 = quarter.normalize(body1.position)
    val p2 = quarter.normalize(body2.position)

    val m1 = body1.mass
    val m2 = body2.mass

    val dx = p2.x - p1.x
    val dy = p2.y - p1.y
    val r = hypotenuse(dx, dy)

    val F = G * m1 * m2 / math.pow(r, 2)
    val a = F / m1

    quarter.normalize(vector(dy / dx, a))
  }
}

File: scarl-ui/src/system/physics.js
import * as coords from './coordinates'
import Vector from './Vector'

export const g = 9.80665

export const G = 6.67408 * Math.pow(10, -11)

export const AU = 149597871000

export const hypotenuse = (a, b) => Math.sqrt(Math.pow(a, 2) + Math.pow(b, 2))

export const vector = (slope, length) => {
    const x = length / Math.sqrt(1 + Math.pow(slope, 2))
    const y = slope * x

    return Vector({x,y})
}

export const travelTime = (distance, acceleration) => Math.sqrt(2 * distance / acceleration)

export const gravity = (bodies, body) => bodies.reduce((result, other) => (
    other.id === body.id ? result : (
        Vector.add(result, calculateGravity(body, other))
    )
), Vector({x: 0, y: 0}))

const calculateGravity = (body1, body2) => {
    const quarter = coords.quarter(body1.position, body2.position)
    const p1 = coords.normalize(quarter, body1.position)
    const p2 = coords.normalize(quarter, body2.position)

    const m1 = body1.mass
    const m2 = body2.mass

    const dx = p2.x - p1.x
    const dy = p2.y - p1.y

    const r = hypotenuse(dx, dy)

    const F = G * m1 * m2 / Math.pow(r, 2)
    const a = F/m1

    return coords.normalize(quarter, vector(dy/dx, a))
}
